File: apple/MuCore/Sources/MuCore/AudioDsp.swift
import Foundation

/// 單一 biquad（RBJ cookbook peaking EQ），a0 已正規化。浮點——不入三方 byte 契約，
/// 由各平台單元測試驗頻率響應（model.md §1.10）。
public struct Biquad: Equatable {
    public let b0: Double, b1: Double, b2: Double, a1: Double, a2: Double

    public init(b0: Double, b1: Double, b2: Double, a1: Double, a2: Double) {
        self.b0 = b0; self.b1 = b1; self.b2 = b2; self.a1 = a1; self.a2 = a2
    }

    public static let identity = Biquad(b0: 1, b1: 0, b2: 0, a1: 0, a2: 0)

    /// f0 ≥ Nyquist 或增益 0 → identity。
    public static func peaking(sampleRate: Double, f0: Double, gainDb: Double,
                               q: Double = EqSettings.bandQ) -> Biquad {
        guard sampleRate > 0, f0 > 0, f0 < sampleRate / 2, gainDb != 0, q > 0 else { return identity }
        let a = pow(10.0, gainDb / 40.0)
        let w0 = 2.0 * Double.pi * f0 / sampleRate
        let alpha = sin(w0) / (2.0 * q)
        let cosW0 = cos(w0)
        let a0 = 1 + alpha / a
        return Biquad(b0: (1 + alpha * a) / a0,
                      b1: (-2 * cosW0) / a0,
                      b2: (1 - alpha * a) / a0,
                      a1: (-2 * cosW0) / a0,
                      a2: (1 - alpha / a) / a0)
    }

    /// |H(e^jw)|（測試用）。
    public func magnitude(atHz f: Double, sampleRate: Double) -> Double {
        let w = 2.0 * Double.pi * f / sampleRate
        let cw = cos(w), sw = sin(w)
        let c2w = cos(2 * w), s2w = sin(2 * w)
        let numRe = b0 + b1 * cw + b2 * c2w
        let numIm = -(b1 * sw + b2 * s2w)
        let denRe = 1 + a1 * cw + a2 * c2w
        let denIm = -(a1 * sw + a2 * s2w)
        return (numRe * numRe + numIm * numIm).squareRoot()
            / (denRe * denRe + denIm * denIm).squareRoot()
    }
}

/// 播放期 DSP：EQ 串接 + 總增益（model.md §1.10）。
/// 直通時零成本；輸出硬性 clamp 到 ±1.0（正增益過大 → 削峰，屬使用者選擇）。
/// 非 thread-safe 的處理迴圈：`configure` 由控制端呼叫、`process` 由音訊執行緒呼叫，
/// 兩者以 `lock` 序列化（configure 頻率極低）。
public final class AudioDsp {

    private let lock = NSLock()
    private var filters: [Biquad] = []
    private var state: [[Double]] = []       // [channel][filter*4]：x1,x2,y1,y2
    private var gain: Float = 1
    private var identity = true
    private var channels = 0
    private var sampleRate: Double = 0
    private var settings: EqSettings = .default
    private var gainMb = 0

    public init() {}

    /// 目前是否直通（呼叫端可據此略過整個處理）。
    public var isIdentity: Bool {
        lock.lock(); defer { lock.unlock() }
        return identity
    }

    /// 控制端（主執行緒）：更新設定與總增益；格式沿用最後一次 `prepare`。
    public func setSettings(_ settings: EqSettings, gainMb: Int) {
        lock.lock(); defer { lock.unlock() }
        self.settings = settings
        self.gainMb = gainMb
        rebuildLocked()
    }

    /// 音訊端：串流格式就緒/改變（tap prepare）。
    public func prepare(sampleRate: Double, channels: Int) {
        lock.lock(); defer { lock.unlock() }
        self.sampleRate = sampleRate
        self.channels = channels
        rebuildLocked()
    }

    /// 一次設定（測試/單純情境）。
    public func configure(settings: EqSettings, gainMb: Int, sampleRate: Double, channels: Int) {
        lock.lock(); defer { lock.unlock() }
        self.settings = settings
        self.gainMb = gainMb
        self.sampleRate = sampleRate
        self.channels = channels
        rebuildLocked()
    }

    private func rebuildLocked() {
        let newFilters = sampleRate > 0
            ? settings.activeBands().map {
                Biquad.peaking(sampleRate: sampleRate, f0: Double($0.hz),
                               gainDb: Double($0.mb) / 100.0)
              }.filter { $0 != .identity }
            : []
        let changedShape = newFilters.count != filters.count || state.count != max(channels, 1)
        filters = newFilters
        gain = EqSettings.linear(mb: gainMb)
        identity = newFilters.isEmpty && gainMb == 0
        if changedShape {
            state = Array(repeating: Array(repeating: 0, count: max(newFilters.count, 1) * 4),
                          count: max(channels, 1))
        }
    }

    public func reset() {
        lock.lock(); defer { lock.unlock() }
        for c in state.indices { for i in state[c].indices { state[c][i] = 0 } }
    }

    /// 交錯（interleaved）緩衝原地處理。
    public func process(_ buf: UnsafeMutablePointer<Float>, frames: Int, channels ch: Int) {
        lock.lock(); defer { lock.unlock() }
        guard !identity, frames > 0, ch > 0 else { return }
        if state.count < ch {
            state = Array(repeating: Array(repeating: 0, count: max(filters.count, 1) * 4),
                          count: ch)
        }
        for c in 0..<ch {
            state[c].withUnsafeMutableBufferPointer { st in
                for f in 0..<frames {
                    var x = Double(buf[f * ch + c])
                    for (k, bq) in filters.enumerated() {
                        x = Self.step(bq, x, st.baseAddress! + k * 4)
                    }
                    buf[f * ch + c] = Self.clip(Float(x) * gain)
                }
            }
        }
    }

    /// 非交錯（每聲道獨立緩衝）原地處理。
    public func process(planar bufs: [UnsafeMutablePointer<Float>], frames: Int) {
        lock.lock(); defer { lock.unlock() }
        guard !identity, frames > 0, !bufs.isEmpty else { return }
        if state.count < bufs.count {
            state = Array(repeating: Array(repeating: 0, count: max(filters.count, 1) * 4),
                          count: bufs.count)
        }
        for (c, buf) in bufs.enumerated() {
            state[c].withUnsafeMutableBufferPointer { st in
                for f in 0..<frames {
                    var x = Double(buf[f])
                    for (k, bq) in filters.enumerated() {
                        x = Self.step(bq, x, st.baseAddress! + k * 4)
                    }
                    buf[f] = Self.clip(Float(x) * gain)
                }
            }
        }
    }

    /// 測試用：交錯陣列處理。
    public func processed(_ samples: [Float], channels ch: Int) -> [Float] {
        var out = samples
        out.withUnsafeMutableBufferPointer { p in
            process(p.baseAddress!, frames: p.count / ch, channels: ch)
        }
        return out
    }

    /// Direct Form I 單步；st = [x1,x2,y1,y2]。
    private static func step(_ bq: Biquad, _ x: Double, _ st: UnsafeMutablePointer<Double>) -> Double {
        let y = bq.b0 * x + bq.b1 * st[0] + bq.b2 * st[1] - bq.a1 * st[2] - bq.a2 * st[3]
        st[1] = st[0]; st[0] = x
        st[3] = st[2]; st[2] = y
        return y
    }

    private static func clip(_ v: Float) -> Float {
        v > 1 ? 1 : (v < -1 ? -1 : v)
    }
}

File: android/core/src/main/kotlin/crate/core/AudioDsp.kt
package crate.core

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * 單一 biquad（RBJ cookbook peaking EQ），a0 已正規化。浮點——不入三方 byte 契約，
 * 由各平台單元測試驗頻率響應（model.md §1.10）。
 */
data class Biquad(
    val b0: Double, val b1: Double, val b2: Double, val a1: Double, val a2: Double,
) {
    companion object {
        val IDENTITY = Biquad(1.0, 0.0, 0.0, 0.0, 0.0)

        /** f0 ≥ Nyquist 或增益 0 → identity。 */
        fun peaking(
            sampleRate: Double,
            f0: Double,
            gainDb: Double,
            q: Double = EqSettings.BAND_Q,
        ): Biquad {
            if (sampleRate <= 0 || f0 <= 0 || f0 >= sampleRate / 2 || gainDb == 0.0 || q <= 0) {
                return IDENTITY
            }
            val a = 10.0.pow(gainDb / 40.0)
            val w0 = 2.0 * PI * f0 / sampleRate
            val alpha = sin(w0) / (2.0 * q)
            val cosW0 = cos(w0)
            val a0 = 1 + alpha / a
            return Biquad(
                (1 + alpha * a) / a0,
                (-2 * cosW0) / a0,
                (1 - alpha * a) / a0,
                (-2 * cosW0) / a0,
                (1 - alpha / a) / a0,
            )
        }
    }

    /** |H(e^jw)|（測試用）。 */
    fun magnitude(atHz: Double, sampleRate: Double): Double {
        val w = 2.0 * PI * atHz / sampleRate
        val cw = cos(w); val sw = sin(w)
        val c2w = cos(2 * w); val s2w = sin(2 * w)
        val numRe = b0 + b1 * cw + b2 * c2w
        val numIm = -(b1 * sw + b2 * s2w)
        val denRe = 1 + a1 * cw + a2 * c2w
        val denIm = -(a1 * sw + a2 * s2w)
        return sqrt(numRe * numRe + numIm * numIm) / sqrt(denRe * denRe + denIm * denIm)
    }
}

/**
 * 播放期 DSP：EQ 串接 + 總增益（model.md §1.10）。
 * 直通時零成本；輸出硬性 clamp 到 ±1.0（正增益過大 → 削峰，屬使用者選擇）。
 * `configure` 由控制端呼叫、`process` 由音訊執行緒呼叫，兩者以 lock 序列化（configure 頻率極低）。
 */
class AudioDsp {

    private val lock = Any()
    private var filters: List<Biquad> = emptyList()
    private var state: Array<DoubleArray> = arrayOf()
    private var gain: Float = 1f
    @Volatile private var identityFlag = true
    private var channels = 0
    private var sampleRate = 0.0
    private var settings: EqSettings = EqSettings.DEFAULT
    private var gainMb = 0

    /** 目前是否直通（呼叫端可據此略過整個處理）。 */
    val isIdentity: Boolean get() = identityFlag

    /** 控制端：更新設定與總增益；格式沿用最後一次 [prepare]。 */
    fun setSettings(settings: EqSettings, gainMb: Int) {
        synchronized(lock) {
            this.settings = settings
            this.gainMb = gainMb
            rebuildLocked()
        }
    }

    /** 音訊端：串流格式就緒/改變。 */
    fun prepare(sampleRate: Double, channels: Int) {
        synchronized(lock) {
            this.sampleRate = sampleRate
            this.channels = channels
            rebuildLocked()
        }
    }

    /** 一次設定（測試/單純情境）。 */
    fun configure(settings: EqSettings, gainMb: Int, sampleRate: Double, channels: Int) {
        synchronized(lock) {
            this.settings = settings
            this.gainMb = gainMb
            this.sampleRate = sampleRate
            this.channels = channels
            rebuildLocked()
        }
    }

    private fun rebuildLocked() {
        val newFilters = if (sampleRate > 0) {
            settings.activeBands()
                .map { Biquad.peaking(sampleRate, it.first.toDouble(), it.second / 100.0) }
                .filter { it != Biquad.IDENTITY }
        } else {
            emptyList()
        }
        val changedShape = newFilters.size != filters.size || state.size != maxOf(channels, 1)
        filters = newFilters
        gain = EqSettings.linear(gainMb)
        identityFlag = newFilters.isEmpty() && gainMb == 0
        if (changedShape) {
            state = Array(maxOf(channels, 1)) { DoubleArray(maxOf(newFilters.size, 1) * 4) }
        }
    }

    fun reset() {
        synchronized(lock) { state.forEach { it.fill(0.0) } }
    }

    /** 交錯（interleaved）緩衝原地處理；[frames] = 每聲道樣本數。 */
    fun process(buf: FloatArray, offset: Int, frames: Int, ch: Int) {
        synchronized(lock) {
            if (identityFlag || frames <= 0 || ch <= 0) return
            if (state.size < ch) {
                state = Array(ch) { DoubleArray(maxOf(filters.size, 1) * 4) }
            }
            val fs = filters
            for (c in 0 until ch) {
                val st = state[c]
                for (f in 0 until frames) {
                    val i = offset + f * ch + c
                    var x = buf[i].toDouble()
                    for (k in fs.indices) x = step(fs[k], x, st, k * 4)
                    buf[i] = clip((x * gain).toFloat())
                }
            }
        }
    }

    /** 測試用：回傳處理後的新陣列。 */
    fun processed(samples: FloatArray, ch: Int): FloatArray {
        val out = samples.copyOf()
        process(out, 0, out.size / ch, ch)
        return out
    }

    /** Direct Form I 單步；st[base..base+3] = x1,x2,y1,y2。 */
    private fun step(bq: Biquad, x: Double, st: DoubleArray, base: Int): Double {
        val y = bq.b0 * x + bq.b1 * st[base] + bq.b2 * st[base + 1] -
            bq.a1 * st[base + 2] - bq.a2 * st[base + 3]
        st[base + 1] = st[base]; st[base] = x
        st[base + 3] = st[base + 2]; st[base + 2] = y
        return y
    }

    private fun clip(v: Float): Float = if (v > 1f) 1f else if (v < -1f) -1f else v
}
